apps/emqx_bridge/src/emqx_bridge_api.erl

```diff
@@ -485,57 +485,22 @@ schema("/bridges_probe") ->
     RequestMeta = #{module => ?MODULE, method => post, path => "/bridges_probe"},
     case emqx_dashboard_swagger:filter_check_request_and_translate_body(Request, RequestMeta) of
         {ok, #{body := #{<<"type">> := ConnType} = Params}} ->
-            case do_probe(ConnType, maps:remove(<<"type">>, Params)) of
+            case emqx_bridge_resource:create_dry_run(ConnType, maps:remove(<<"type">>, Params)) of
                 ok ->
                     {204};
                 {error, Error} ->
-                    {400, error_msg('TEST_FAILED', Error)}
+                    {400, error_msg('TEST_FAILED', try_json_encode(Error))}
             end;
         BadRequest ->
             BadRequest
     end.
 
-do_probe(ConnType, Params) ->
-    case test_connection(host_and_port(ConnType, Params)) of
-        ok ->
-            emqx_bridge_resource:create_dry_run(ConnType, Params);
-        Error ->
-            Error
-    end.
-
-host_and_port(mqtt, #{<<"server">> := Server}) ->
-    case string:split(Server, ":") of
-        [Host, Port] -> {Host, list_to_integer(Port)};
-        _Other -> error(invalid_server, Server)
-    end;
-host_and_port(webhook, #{<<"url">> := Url}) ->
-    {BaseUrl, _Path} = parse_url(Url),
-    {ok, #{host := Host, port := Port}} = emqx_http_lib:uri_parse(BaseUrl),
-    {Host, Port};
-host_and_port(_Unknown, _) ->
-    undefined.
-
-%% [TODO] remove in EMQX-8588 when resource manager handles things more elegantly
-test_connection(undefined) ->
-    %% be friendly, it might fail later on with a 'timeout' error.
-    ok;
-test_connection({Host, Port}) ->
-    case gen_tcp:connect(Host, Port, [], 5000) of
-        {ok, TestSocket} -> gen_tcp:close(TestSocket);
-        Error -> Error
-    end.
-
-parse_url(Url) ->
-    case string:split(Url, "//", leading) of
-        [Scheme, UrlRem] ->
-            case string:split(UrlRem, "/", leading) of
-                [HostPort, Path] ->
-                    {iolist_to_binary([Scheme, "//", HostPort]), Path};
-                [HostPort] ->
-                    {iolist_to_binary([Scheme, "//", HostPort]), <<>>}
-            end;
-        [Url] ->
-            error({invalid_url, Url})
+try_json_encode(Term) ->
```

Comment:
It seems like error_msg/2 calls emqx:readable_error_msg/1 to try to make the Error term into a readable string. Why does it need to be json encoded?

Reply:
You're totally right, making it as human readable as possible sure is the way to go.

Reply:
The problem is that for all the cases I tested against `emqx_misc:readable_error_msg/1` apparently didn't produce something meaningful and instead just returned a representation of the erlang data structure. Which is less readable than JSON. I thought. But now we'd always have JSON without even trying. Maybe the wrong place to fix this. I'll stick to what we do in general. But to address that in a more generic way.